napari/plugins/_hook_callers.py

```diff
@@ -177,51 +165,321 @@ def _multicall(
+            If ``plugin_name`` has not provided a hook implementation for this
+            hook specification.
+        """
+        self.get_plugin_implementation(plugin_name).enabled = enabled
+
+    def enable_plugin(self, plugin_name: str):
+        """enable implementation for ``plugin_name``."""
+        self._set_plugin_enabled(plugin_name, True)
+
+    def disable_plugin(self, plugin_name: str):
+        """disable implementation for ``plugin_name``."""
+        self._set_plugin_enabled(plugin_name, False)
+
+    def _call_plugin(self, plugin_name: str, *args, **kwargs):
+        """Call the hook implementation for a specific plugin
+
+        Note: this method is not intended to be called directly. Instead, just
+        call the instance directly, specifing the ``_plugin`` argument.
+        See the ``__call__`` method below.
+
+        Parameters
+        ----------
+        plugin_name : str
+            Name of the plugin
+
+        Returns
+        -------
+        Any
+            Result of implementation call provided by plugin
+
+        Raises
+        ------
+        TypeError
+            If the implementation is a hook wrapper (cannot be called directly)
+        TypeError
+            If positional arguments are provided
+        HookCallError
+            If one of the required arguments in the hook specification is not
+            present in ``kwargs``.
+        PluginCallError
+            If an exception is raised when calling the plugin
+        """
+        implementation = self.get_plugin_implementation(plugin_name)
+        if implementation.hookwrapper:
+            raise TypeError("Hook wrappers can not be called directly")
+
+        # pluggy only allows calling hooks with keyword arguments
+        if args:
+            raise TypeError("hook calling supports only keyword arguments")
+        _args: List[Any] = []
+        # this converts kwargs into positional arguments in the correct order
+        # for the hookspec
+        try:
+            _args = [kwargs[argname] for argname in implementation.argnames]
+        except KeyError:
+            for argname in implementation.argnames:
+                if argname not in kwargs:
+                    raise HookCallError(
+                        f"hook call must provide argument {argname}"
+                    )
+
+        try:
+            return implementation.function(*_args)
+        except Exception as exc:
+            raise PluginCallError(
+                implementation.plugin_name,
+                implementation.plugin.__name__,
+                msg=(
+                    f"Error calling plugin '{implementation.plugin_name}', "
+                    f"hook '{str(implementation.function.__name__)}'"
+                ),
+            ) from exc
+
+    def __call__(
+        self,
+        *,
+        _skip_impls: Optional[Sequence[HookImpl]] = None,
+        _return_impl: bool = False,
+        _plugin: Optional[str] = None,
+        **kwargs,
+    ):
```

Comment:
I'm not in love with functions changing the return type based on kwargs. @tlambert03 and discussed on Zulip and came up with:

- `.__call__` -> return only the result
- `.call_with_impl` -> return an object containing the result, the implementation used, and other metadata.

This echoes some discussions we had at scikit-image e.g. [here](https://github.com/scikit-image/scikit-image/issues/3892#issuecomment-591819705) and in these [meeting notes](https://github.com/scikit-image/meeting-notes/blob/master/2020/2020-02-27--BIDS-core-dev-meeting.markdown) (search for "API policies")

Reply:
`_skip_impl` and `_plugin` are fine.

Reply:
Also: I wonder whether we want to support returning of a list. This was a pluggy thing, but I don't know whether we will have hookspecs for which we want to call *all* implementations. What do you think @tlambert03? (Just trying to simplify the return type to always be consistent...)